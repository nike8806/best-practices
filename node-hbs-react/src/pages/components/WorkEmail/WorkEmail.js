import React, { Component } from 'react';
import PropTypes from 'prop-types';


class WorkEmail extends Component {
  static propTypes = {
    onSubmit: PropTypes.func.isRequired
  };

  state = {
    email: null,
    validated: false,
    isValid: false,
    emailErrorMessage: null
  };

  /**
   * Sync state with the user input
   *
   * @param {SyntheticEvent}
   */
  handleInput = (event) => {
    const email = event.target.value;

    this.setState(() => ({
      email,
      isValid: this._isValid(email)
    }));
  }

  /**
   * Catch the attempt to submit
   *
   * @param {SyntheticEvent}
   */
  handleSubmit = (event) => {
    this.setState(() => ({
      validated: true,
      isValid: this._isValid(this.state.email)
    }));

    if (this._isValid(this.state.email)) {
      this.props.onSubmit(this.state.email);
    }
  }

  /**
   * Validate if the email is not from a public service. It's the condition to validate as a work email.
   *
   * @param {string} Work email to validate
   * @return {boolean}
   */
  _isValid(email) {
    let condition = new RegExp('^[A-Za-z0-9._%+-]+@(?:[A-Za-z0-9-]+\\.)+[A-Za-z]{2,4}$');

    // Check if is valid email
    if (!condition.test(email)) {
      this.setState({
        emailErrorMessage: 'Please specify a valid email address'
      });
      return false;
    }

    // Invalidate email as work email for listed domains. See ticket UWOPS-141 for details
    condition = /^[A-Za-z0-9._%+-]+@(?!(?:gmail|hotmail|yahoo|aol|comcast|msn))\w+[-\.\w]*?\.\w{2,4}$/;
    if (!condition.test(email)) {
      this.setState({
        emailErrorMessage: 'Sorry, we can\'t accept addresses from popular email providers (gmail, hotmail, yahoo, etc.). If you don\'t have any other work email address, please select \'I don\'t have a work email.\''
      });
      return false;
    }
    this.setState({
      emailErrorMessage: null
    });

    return true;
  }

  /**
   * Render method
   *
   * @return {ReactElement}
   */
  render() {
    const errorClass = (this.state.validated && !this.state.isValid) ? 'has-error' : '';
    const emailErrorMessage = this.state.emailErrorMessage;

    return (
      <div className={ errorClass }>
        <input type="email" className="form-control" onChange={ this.handleInput } />
        { !errorClass ? null :
          <div className="error-message">
            <span className="icon icon-form-error error-image"></span>
            <span className="error-text">{emailErrorMessage}</span>
          </div>
        }
        <button type="submit" className="btn btn--blue work-email-confirmation__btn-submit-email" onClick={ this.handleSubmit }>
          Submit
        </button>
      </div>
    );
  }
}

export default WorkEmail;
