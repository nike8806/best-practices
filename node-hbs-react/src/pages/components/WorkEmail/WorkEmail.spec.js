import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import WorkEmail from './WorkEmail';

// TODO: Put this configuration in just one place
import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({ adapter: new Adapter() });
const cbSubmitWorkEmailMock = () => { };

test('Input Work Email render correctly', () => {
  const component = shallow(<WorkEmail onSubmit={cbSubmitWorkEmailMock} />);

  expect(component).toMatchSnapshot();
});

test('Capture the user input', () => {
  const component = shallow(<WorkEmail onSubmit={cbSubmitWorkEmailMock} />);

  component.find('input').simulate('change', { target: { value: 'some@email.tld' } });

  expect(component.state().email).toBe('some@email.tld');
});

test('Valite when click the button', () => {
  const component = shallow(<WorkEmail onSubmit={cbSubmitWorkEmailMock} />);

  expect(component.state().validated).toBe(false);

  component.find('button').simulate('click');
  expect(component.state().validated).toBe(true);
});

test('Display error message when try submit without email', () => {
  const component = shallow(<WorkEmail onSubmit={cbSubmitWorkEmailMock} />);

  expect(component.find('.error-message').length).toBe(0);

  component.find('button').simulate('click');
  expect(component.update().find('.error-message').length).toBe(1);
  expect(component).toMatchSnapshot();
});

test('Display error message when try submit an invalid email', () => {
  const component = shallow(<WorkEmail onSubmit={cbSubmitWorkEmailMock} />);

  expect(component.find('.error-message').length).toBe(0);

  component.find('input').simulate('change', { target: { value: 'user@gmail.com' } });
  component.find('button').simulate('click');

  component.update();

  expect(component.state().email).toBe('user@gmail.com');
  expect(component.find('.error-message').length).toBe(1);
  expect(component).toMatchSnapshot();
});
