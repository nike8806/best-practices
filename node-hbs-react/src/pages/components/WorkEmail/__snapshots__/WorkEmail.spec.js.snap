// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display error message when try submit an invalid email 1`] = `
<div
  className="has-error"
>
  <input
    className="form-control"
    onChange={[Function]}
    type="email"
  />
  <div
    className="error-message"
  >
    <span
      className="icon icon-form-error error-image"
    />
    <span
      className="error-text"
    >
      Sorry, we can't accept addresses from popular email providers (gmail, hotmail, yahoo, etc.). If you don't have any other work email address, please select 'I don't have a work email.'
    </span>
  </div>
  <button
    className="btn btn--blue work-email-confirmation__btn-submit-email"
    onClick={[Function]}
    type="submit"
  >
    Submit
  </button>
</div>
`;

exports[`Display error message when try submit without email 1`] = `
<div
  className="has-error"
>
  <input
    className="form-control"
    onChange={[Function]}
    type="email"
  />
  <div
    className="error-message"
  >
    <span
      className="icon icon-form-error error-image"
    />
    <span
      className="error-text"
    >
      Please specify a valid email address
    </span>
  </div>
  <button
    className="btn btn--blue work-email-confirmation__btn-submit-email"
    onClick={[Function]}
    type="submit"
  >
    Submit
  </button>
</div>
`;

exports[`Input Work Email render correctly 1`] = `
<div
  className=""
>
  <input
    className="form-control"
    onChange={[Function]}
    type="email"
  />
  <button
    className="btn btn--blue work-email-confirmation__btn-submit-email"
    onClick={[Function]}
    type="submit"
  >
    Submit
  </button>
</div>
`;
